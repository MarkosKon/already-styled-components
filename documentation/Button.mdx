---
name: Button
menu: Components
---

import { Playground, PropsTable } from "docz";
import Button from "../src/components/Button";

# Button

## Basic Usage

<Playground>
  <Button>Click me!</Button>
  <Button bc="rebeccapurple" c="white" hc="yellow">
    Color props
  </Button>
  <Button transparent c="black" fs="32px">Icon button</Button>
</Playground>

## Props

<PropsTable of={Button} />
