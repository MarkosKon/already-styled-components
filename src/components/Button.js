import PropTypes from "prop-types";
import styled from "styled-components"

const Button = styled.button`
  color: #fff;
  background-color: ${({ transparent }) =>
    transparent ? "transparent" : "#00AFB1"};
  font-family: 'Inconsolata', monospace;
  font-size: ${({ fontSize }) => fontSize};
  font-weight: bold;
  padding: 15px 30px;
  border: none;
  border-radius: 20px;
  cursor: pointer;
  ${({ transparent }) =>
    !transparent &&
    `
    transition: background-color 0.3s ease-out;
    &:hover {
        background-color: #038486;
    }
  `};
`;

Button.propTypes = {
  transparent: PropTypes.bool,
  fontSize: PropTypes.string,
  onClick: PropTypes.func
};

Button.defaultProps = {
  transparent: false,
  fontSize: "20px"
};

export default Button;