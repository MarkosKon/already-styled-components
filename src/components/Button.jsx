import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  getLuminance, darken, lighten, transparentize,
} from 'polished';

const StyledButton = styled.button`
  color: ${({ c }) => c};
  background-color: ${({ transparent, bc }) => (transparent ? 'transparent' : bc)};
  font-family: ${({ ff }) => ff};
  font-size: ${({ fs }) => fs};
  font-weight: bold;
  padding: 15px 30px;
  border: none;
  border-radius: 20px;
  outline: 0;
  cursor: pointer;
  transition: color 0.3s ease-out, background-color 0.3s ease-out;
  ${({ hc }) => hc
    && `
    &:hover {
      color: ${hc};
    }
  `}
  ${({ transparent, bc }) => !transparent
    && `
    &:hover {
        background-color: ${getLuminance(bc) < 0.06 ? lighten(0.15, bc) : darken(0.07, bc)};
    }
    &:focus {
        box-shadow: 0 0 0 0.2rem ${
  getLuminance(bc) > 0.8 ? darken(0.15, bc) : transparentize(0.7, bc)
};
    }
  `};
`;

const Button = ({
  transparent = false,
  className = null,
  children = null,
  c = '#FFF',
  bc = '#00AFB1',
  hc = null,
  ff = 'inherit',
  fs = '20px',
}) => (
  <StyledButton
    transparent={transparent}
    className={className}
    c={c}
    bc={bc}
    hc={hc}
    ff={ff}
    fs={fs}
  >
    {children}
  </StyledButton>
);

if (process.env.NODE_ENV !== 'production') {
  Button.propTypes = {
    transparent: PropTypes.bool,
    className: PropTypes.string,
    children: PropTypes.node,
    c: PropTypes.string,
    bc: PropTypes.string,
    hc: PropTypes.string,
    ff: PropTypes.string,
    fs: PropTypes.string,
  };
  Button.defaultProps = {
    transparent: false,
    className: null,
    children: null,
    c: '#FFF',
    bc: '#00AFB1',
    hc: null,
    ff: 'inherit',
    fs: '20px',
  };
}

export default Button;
