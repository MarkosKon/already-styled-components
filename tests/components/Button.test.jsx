import React from 'react';
import renderer from 'react-test-renderer';
import {
  getLuminance, darken, lighten, transparentize,
} from 'polished';
import 'jest-styled-components';

import Button from '../../src/components/Button';

test('Button #1 - Default props.', () => {
  const tree = renderer.create(<Button />).toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('color', '#FFF');
  expect(tree).toHaveStyleRule('background-color', '#00AFB1');
  expect(tree).toHaveStyleRule('border-radius', '20px');
  expect(tree).toHaveStyleRule('font-family', 'inherit');
  expect(tree).toHaveStyleRule('font-size', '20px');
  expect(tree).not.toHaveStyleRule('color', {
    modifier: '&:hover',
  });
  expect(tree).toHaveStyleRule('background-color', '#008c8d', {
    modifier: '&:hover',
  });
  expect(tree).toHaveStyleRule('box-shadow', '0 0 0 0.2rem rgba(0,175,177,0.3)', {
    modifier: '&:focus',
  });
});

test('Button #2 - All props present.', () => {
  const tree = renderer
    .create(<Button transparent c="yellow" bc="black" hc="blue" br="0" ff="Arial" fs="32px" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('color', 'yellow');
  expect(tree).toHaveStyleRule('background-color', 'transparent');
  expect(tree).toHaveStyleRule('border-radius', '0');
  expect(tree).toHaveStyleRule('font-family', 'Arial');
  expect(tree).toHaveStyleRule('font-size', '32px');
  expect(tree).toHaveStyleRule('color', 'blue', {
    modifier: '&:hover',
  });
  expect(tree).not.toHaveStyleRule('background-color', {
    modifier: '&:hover',
  });
  expect(tree).not.toHaveStyleRule('box-shadow', {
    modifier: '&:focus',
  });
});

test('Button #3 - Background-color, hover-color.', () => {
  const tree = renderer.create(<Button bc="azure" hc="yellow" />).toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('color', '#FFF');
  expect(tree).toHaveStyleRule('background-color', 'azure');
  expect(tree).toHaveStyleRule('color', 'yellow', {
    modifier: '&:hover',
  });
  expect(tree).toHaveStyleRule(
    'background-color',
    `${getLuminance('azure') < 0.06 ? lighten(0.15, 'azure') : darken(0.07, 'azure')}`,
    {
      modifier: '&:hover',
    },
  );
  expect(tree).toHaveStyleRule(
    'box-shadow',
    `0 0 0 0.2rem ${
      getLuminance('azure') > 0.8 ? darken(0.15, 'azure') : transparentize(0.7, 'azure')
    }`,
    {
      modifier: '&:focus',
    },
  );
});
