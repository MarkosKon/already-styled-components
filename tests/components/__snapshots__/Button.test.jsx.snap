// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button #1 - Default props. 1`] = `
.c0 {
  color: #FFF;
  background-color: #00AFB1;
  font-family: inherit;
  font-size: 20px;
  font-weight: bold;
  padding: 15px 30px;
  border: none;
  border-radius: 20px;
  outline: 0;
  cursor: pointer;
  -webkit-transition: color 0.3s ease-out,background-color 0.3s ease-out;
  transition: color 0.3s ease-out,background-color 0.3s ease-out;
}

.c0:hover {
  background-color: #008c8d;
}

.c0:focus {
  box-shadow: 0 0 0 0.2rem rgba(0,175,177,0.3);
}

<button
  className="c0"
/>
`;

exports[`Button #2 - All props present. 1`] = `
.c0 {
  color: yellow;
  background-color: transparent;
  font-family: Arial;
  font-size: 32px;
  font-weight: bold;
  padding: 15px 30px;
  border: none;
  border-radius: 0;
  outline: 0;
  cursor: pointer;
  -webkit-transition: color 0.3s ease-out,background-color 0.3s ease-out;
  transition: color 0.3s ease-out,background-color 0.3s ease-out;
}

.c0:hover {
  color: blue;
}

<button
  className="c0"
/>
`;

exports[`Button #3 - Background-color, hover-color. 1`] = `
.c0 {
  color: #FFF;
  background-color: azure;
  font-family: inherit;
  font-size: 20px;
  font-weight: bold;
  padding: 15px 30px;
  border: none;
  border-radius: 20px;
  outline: 0;
  cursor: pointer;
  -webkit-transition: color 0.3s ease-out,background-color 0.3s ease-out;
  transition: color 0.3s ease-out,background-color 0.3s ease-out;
}

.c0:hover {
  color: yellow;
}

.c0:hover {
  background-color: #cff;
}

.c0:focus {
  box-shadow: 0 0 0 0.2rem #a3ffff;
}

<button
  className="c0"
/>
`;
